storage/src/store/mod.rs

```diff
@@ -0,0 +1,619 @@
+
+    /// The codec configuration to use for encoding and decoding log items.
+    pub log_codec_config: C,
+
+    /// The number of operations to store in each section of the [`Journal`].
+    pub log_items_per_section: u64,
+
+    /// The [`Translator`] used by the compressed index.
+    pub translator: T,
+}
+
+/// An unauthenticated key-value database based off of an append-only [`Journal`] of operations.
+pub struct Store<E, K, V, T>
+where
+    E: RStorage + Clock + Metrics,
+    K: Array,
+    V: Span,
+    T: Translator,
+{
+    /// A log of all [`Variable`] operations that have been applied to the store.
+    pub(super) log: Journal<E, Variable<K, V>>,
+
+    /// A snapshot of all currently active operations in the form of a map from each key to the
+    /// section and offset within the section containing its most recent update.
+    ///
+    /// # Invariant
+    ///
+    /// Only references operations of type [`Variable::Set`].
+    pub(super) snapshot: Index<T, SectionAndOffset>,
+
+    /// The number of items to store in each section of the variable journal.
+    pub(super) log_items_per_section: u64,
+
+    /// The total numer of operations that have been applied to the store.
+    pub(super) op_count: u64,
+
+    /// The number of operations that are pending commit.
+    pub(super) uncommitted_ops: u64,
+}
+
+impl<E, K, V, T> Store<E, K, V, T>
+where
+    E: RStorage + Clock + Metrics,
+    K: Array,
+    V: Span,
+    T: Translator,
+{
+    /// Initializes a new [`Store`] database with the given configuration.
+    ///
+    /// ## Rollback
+    ///
+    /// Any uncommitted operations will be rolled back if the [`Store`] was previously closed without
+    /// committing.
+    pub async fn init(
+        context: E,
+        cfg: Config<T, <Variable<K, V> as Read>::Cfg>,
+    ) -> Result<Self, Error> {
+        let mut snapshot: Index<T, SectionAndOffset> =
+            Index::init(context.with_label("snapshot"), cfg.translator);
+
+        let mut log = Journal::init(
+            context.with_label("log"),
+            JConfig {
+                partition: cfg.log_journal_partition,
+                compression: cfg.log_compression,
+                codec_config: cfg.log_codec_config,
+                write_buffer: cfg.log_write_buffer,
+            },
+        )
+        .await?;
+
+        // Build the snapshot by replaying the log of operations.
+        let op_count = Self::build_snapshot_from_log(&mut log, &mut snapshot).await?;
+
+        Ok(Self {
+            log,
+            snapshot,
+            log_items_per_section: cfg.log_items_per_section,
+            op_count,
+            uncommitted_ops: 0,
+        })
+    }
+
+    /// Gets the current value associated with the given key in the store.
+    ///
+    /// If the key does not exist, returns `Ok(None)`.
+    pub async fn get_current(&self, key: &K) -> Result<Option<V>, Error> {
+        self.get_historical(key, self.get_value_count(key).saturating_sub(1))
```

Comment:
As discussed in the comment on `get_historical`, this is not guaranteed to return the current value of a key (if it exists).